cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -222,12 +230,26 @@ protected void triggerShutdown() {
   }
 
   private void persistRecordsMessageIds(Deque<MetricValues> metricValues, Map<TopicIdMetaKey, byte[]> messageIds) {
+    if (metricValues.isEmpty()) {
+      return;
+    }
     try {
-      if (!metricValues.isEmpty()) {
-        persistRecords(metricValues);
-      }
+      long now = System.currentTimeMillis();
+      long lastRecordTime = metricValues.peekLast().getTimestamp();
+      long delay = now - TimeUnit.SECONDS.toMillis(lastRecordTime);
+      metricValues.add(
+        new MetricValues(metricsContextMap, TimeUnit.MILLISECONDS.toSeconds(now),
+                         ImmutableList.of(
+                           new MetricValue("metrics.process.count", MetricType.COUNTER, metricValues.size()),
+                           new MetricValue("metrics.process.delay.ms", MetricType.GAUGE, delay))));
+      metricStore.add(metricValues);
+      recordsProcessed += metricValues.size();
+      PROGRESS_LOG.debug("{} metrics records persisted. Last metric record's timestamp: {}. " +
+                           "Metrics process delay: {}ms", recordsProcessed, lastRecordTime, delay);
       try {
-        metaTable.saveMessageIds(messageIds);
+        if (!messageIds.isEmpty()) {
```

Comment:
This can happen if exception is thrown but not caught when adding metrics to `records` while `messageIds` is still empty, and `persistRecordsMessageIds` is called at the end of the `run()` method